### ISSUE TITLE
[Bug]: Using a class instance with a `reduce` intent doesn't resolve

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#24886]
### Summary of Problem

Using a class instance with a `reduce` intent doesn't resolve. I think we need to make some adjustments in the reduction implementation.

**Description:**
```
$CHPL_HOME/modules/internal/ChapelReduce.chpl:111: In function 'chpl__sumType':
$CHPL_HOME/modules/internal/ChapelReduce.chpl:116: error: cannot default-initialize x: owned MyType
$CHPL_HOME/modules/internal/ChapelReduce.chpl:130: error: use here prevents split-init
note: non-nilable class type 'borrowed MyType' does not support default initialization
note: Consider using the type managed _owned(MyType)? instead
  $CHPL_HOME/modules/internal/ChapelReduce.chpl:142: called as chpl__sumType(type eltType = owned MyType) from initializer
  $CHPL_HOME/test/gpu/native/reduction/degenAccumulate.chpl:17: called as SumReduceScanOp.init(l: chpl_LocalSpinlock, eltType: owned MyType, value: int(64))
```

**Is this a blocking issue with no known work-arounds?**
no

### Steps to Reproduce

**Source Code:**

```chapel
class MyType {
  var x: int;
}

operator MyType.+(first, second) {
  return first.x + second.x;
}

var obj = new MyType();

forall i in 1..10 with (+ reduce obj) {
  obj += new MyType(i);
}

writeln(obj);
```

**Compile command:**
`chpl foo.chpl`

### Configuration Information

Shouldn't matter.

**Comment by DanilaFe** [comment]
I've always wondered where the "default initialization" / identity element requirement comes in. Is it to avoid switching communication patterns depending on if each distributed part of the loop was empty / non-empty?

It seems like any (commutative) semigroup should be reduceable using the group operation -- no identity element needed.

**Comment by bradcray** [comment]
> I've always wondered where the "default initialization" / identity element requirement comes in.

I think you've got the right model in mind.  I tend to consider it to be due to two things (which are really one thing from two perspectives):

* What value should `result` have in:
  ```chapel
  var A: [1..0] real;  // note that this is an empty / size-0 array
  var result = myReduceOp reduce A;
  ```

* In:
  ```chapel
  var result = 42;
  forall i in 1..1000 with (myReduceOp reduce result) {
    result reduce= f(i);
  }
  ```
  What value should each task-private copy of `result` be initialized to?

> It seems like any (commutative) semigroup should be reduceable using the group operation -- no identity element needed.

I don't remember my math classes well enough to understand what you're saying here or see what path forward we could take for the above cases if it's suggesting one.

**Comment by DanilaFe** [comment]
```Chapel
forall i in 1..1000 with (myReduceOp reduce result) {
  result reduce= f(i);
}
```

Hmm. I was actually imagining that `result` would get non-zero-initialized in that case (each serial portion of the forall would receive an initial value of `42`, and the sum would end up counting 42 that many more times. So I didn't think we needed an identity element). But it sounds like the initial value of a reduce variable is the identity element? That would explain it.

Of course, I hadn't considered the case of reduce expressions (where you don't have an initial value). That said, I would not be opposed to requiring an identity element for them, but not for `forall`s with reduce intents -- if only we didn't zero-initialize the accumulator variable.

**Comment by bradcray** [comment]
> I was actually imagining that result would get non-zero-initialized in that case (each serial portion of the forall would receive an initial value of 42, and the sum would end up counting 42 that many more times.

No, users have made it pretty clear that they wouldn't want that behavior, and I think is completely reasonable.  E.g., today, we can write something like:

```chapel
var total = 0;
forall a in A with (+ reduce total) do
  total += a;
forall b in B with (+ reduce total) do
  total += b;
forall c in C with (+ reduce total) do
  total += c;
```

and get a reasonable result.  Whereas if each task carried forth the incoming value of `total`, things would explode.

Nor do they want `result` to be 0 if they did:

```chapel
var result = 42;
forall i in 1..0 with (+ reduce result) {  // note that this is an empty forall loop
  result reduce= f(i);
}
```

(we did one of those two things originally—probably the latter—and there were protests.  I _think_ what we've ended up here matches OpenMP, but am not confident about that).

> But it sounds like the initial value of a reduce variable is the identity element?

Right, because the reduction accumulates elements into the running state.  So if you think of there as being an input type, state type, and output type in the most general case of a reduction, the pattern for any task is something like:

```chapel
var state: stateType = myReduce.ident();

for elem in elementsBeingReduced do
  myReduce.accumulate(state, elem);
```

and then states from tasks are combined using something like:

```chapel
myReduce.combine(myState, anotherState);
```

and then the output value is generated as a final step:

```chapel
var result = myReduce.generate();
```

(I probably don't have these interfaces or names quite right, just sketching the pattern).

> if only we didn't zero-initialize the accumulator variable.

You may just be writing that from the sum reduction's perspective, but to make sure:  we don't zero-initialize the state since, for a `* reduce` or `min reduce`, you wouldn't want it to start as `0`.  It's identity-initialized.

**Comment by DanilaFe** [comment]
> You may just be writing that from the sum reduction's perspective, but to make sure: we don't zero-initialize the state since, for a * reduce or min reduce, you wouldn't want it to start as 0. It's identity-initialized.

Yes, right, I understand. I did mean it for summing specifically and definitely understand that it's the identity element. 

As for the rest of your explanation: it makes sense, and even as I was writing my previous comment I realized that using the identity element per local copy of the variable makes by far the most sense. So I can't think of anything to be done here. 

On the whole, then, the bug in the OP sounds like... Not a bug. If you need a zero element, it should be possible to construct one. Owned classes don't have one (you can't default initialize an owned variable). On the other hand, nilable owned classes do... 

My knee jerk reaction to this is: what if we had a "pragma last resort" definition of operator + on nilable class types, that's defined if the (owned, non-nilable) addition is defined? I'm imagining the semantics as:

```
nil + c = c
c + nil = c
c1 + c2 = (result of involving the non-nilable owned version of +)
```

Then, all non-nil values of the intermediate accumulator can be combined with `obj`. This would require some extra logic in the sum reduction's implementation, but it's not impossible I suppose.
